Roma/roma.1
.TH ROMA 1 "13th May 2013" "TEI .."
.SH NAME
roma \- compile a TEI ODD specification into schemas and dtd
.SH SYNOPSIS
.B roma
.RI [ options ]
.I oddfile
.I outputdir
.\"=====================================================================
.SH DESCRIPTION
This manual page is not meant to be exhaustive.  There
is some more information in the help files for the web version
of Roma, and plenty of background data about ODDs in the 
TEI Guidelines. roma --help may give a more exhaustive guide 
to using the command-line tool.
.PP
.B roma
is used to generate XML RelaxNG or W3C schemas and DTDs 
from the the ODD source in
.IR oddfile
and (optionally) some documentation to go with them.
.\"=====================================================================
.SH OPTIONS
.B roma
accepts the following options:
.TP
.BI --xsl= \ URL
location of TEI XSLT stylesheets, defaulting to /usr/share/xml/tei/stylesheet
.TP
.BI --teiserver= \ URL
location of TEI eXist database server, defaulting to http://tei.oucs.ox.ac.uk/Query/
.TP
.BI --doc         
create expanded documented ODD
.TP
.BI --nodtd       
suppress DTD creation
.TP
.BI --norelax     
suppress RelaxNG creation
.TP
.BI --noxsd       
suppress W3C XML Schema creation
.TP
.BI --debug       
leave temporary files, etc.
.TP
.BI --localsource=
local copy of P5 sources. Run make in P5 source directory to generate p5subset.xml, then point localsource at it.
.TP
.BI --compile         
create compiled file odd
.TP
.BI --doc             
create expanded documented ODD (TEI Lite XML)
.TP
.BI --dochtml         
create HTML version of doc
.TP
.BI --doclang=LANG    
language for documentation
.TP
.BI --docpdf          
create PDF version of doc
.TP
.BI --lang=LANG       
language for names of attributes and elements
.TP
.BI --autoglobal      
adds global attribute class to all elements
.TP
.BI --schematron      
extract Schematron rules
.TP
.BI --isoschematron   
extract ISO Schematron rules
.TP
.BI --useteiversion   
use version data from TEI P5
.TP
.BI --parameterize    
create parameterized DTD
.TP
.BI --patternprefix=STRING
prefix RELAX NG patterns with STRING
.TP
.BI --schema=NAME     
select named schema spec
.TP
.BI --version         
which version of Roma is this

.\"=====================================================================
.SH BUGS
If the ODD specification is internally inconsistent (eg some element has
been deleted on which another element depends), then you may get
a Relax NG schema which is apparently OK, but which cannot be
translated into W3C schema. Because the conversions to DTD and Relax
are maintained separately, some constructs may cause one to fail but
not the other.
No attempt is made to trap error messages from 
.I trang
which is used to make W3C schemas.
.\"=====================================================================
.SH "SEE ALSO"
.BR trang (1)


ChangeLog
2006-04-16  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   M /trunk/P5/ChangeLog
	   M /trunk/gnuify-changelog.pl

	update changelog

2006-04-16  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   M /trunk/P5/odd2htmlp5.xsl
	   M /trunk/P5/odd2htmlp5.xsl.model
	   M /trunk/Stylesheets/html/core.xsl
	   M /trunk/Stylesheets/html/tagdocs.xsl
	   M /trunk/Stylesheets/html/textstructure.xsl
	   M /trunk/Stylesheets/odds/subsetGuidelines.xsl
	   M /trunk/Stylesheets/slides/teihtml-slides.xsl
	   M /trunk/Stylesheets/tei.css
	   M /trunk/Stylesheets/teic.css
	   M /trunk/Stylesheets/teislides.css

	fixes to not use <pre>, but instead <div class="pre">,
	as a <pre> within a <p> is not valid HTML

2006-04-16  Lou Burnard <lou.burnard@oucs.ox.ac.uk>
	Changed paths:
	   M /trunk/P5/Test/testnames.xml

	fix language codes

2006-04-16  Syd Bauman  <sbauman@brown.edu>
	Changed paths:
	   M /trunk/P5/Test/Makefile

	Changed path to main guidelines driver file in Test/Makefile.

2006-03-23  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   M /trunk/P5/schemas.xml
	   D /trunk/P5/unusedSource

	housekeeping

2006-03-23  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   D /trunk/P5/Source/Guidelines/fr/x

	evil rubbish file

2006-03-23  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   D /trunk/P5/Source/BIB
	   D /trunk/P5/Source/COL
	   D /trunk/P5/Source/DEDICATION
	   D /trunk/P5/Source/FM1
	   D /trunk/P5/Source/Guidelines/Odds/.xml
	   D /trunk/P5/Source/PARTIND
	   D /trunk/P5/Source/PREFS
	   D /trunk/P5/Source/REFCLA
	   D /trunk/P5/Source/REFENT
	   D /trunk/P5/Source/REFTAG
	   D /trunk/P5/Source-driver.xml
	   D /trunk/P5/ents.dtd
	   D /trunk/P5/fileents.dtd

	delete unwanted files and directories

2006-03-23  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   D /trunk/P5/Source/AI
	   D /trunk/P5/Source/CC
	   D /trunk/P5/Source/CE
	   D /trunk/P5/Source/CF
	   D /trunk/P5/Source/CH
	   D /trunk/P5/Source/CO
	   D /trunk/P5/Source/DI
	   D /trunk/P5/Source/DR
	   D /trunk/P5/Source/DS
	   D /trunk/P5/Source/DT
	   D /trunk/P5/Source/FD
	   D /trunk/P5/Source/FS
	   D /trunk/P5/Source/FT
	   D /trunk/P5/Source/GD
	   D /trunk/P5/Source/HD
	   D /trunk/P5/Source/IN
	   D /trunk/P5/Source/MD
	   D /trunk/P5/Source/MS
	   D /trunk/P5/Source/ND
	   D /trunk/P5/Source/NH
	   D /trunk/P5/Source/PH
	   D /trunk/P5/Source/SA
	   D /trunk/P5/Source/SG
	   D /trunk/P5/Source/SH
	   D /trunk/P5/Source/ST
	   D /trunk/P5/Source/TC
	   D /trunk/P5/Source/TD
	   D /trunk/P5/Source/TE
	   D /trunk/P5/Source/TS
	   D /trunk/P5/Source/VE
	   D /trunk/P5/Source/WD

	delete directories

2006-03-23  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   D /trunk/P5/Source/AB

	die, directory

2006-03-23  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   A /trunk/P5/Source/Guidelines/fr/AB.xml
	   A /trunk/P5/Source/Guidelines/fr/AI.xml
	   A /trunk/P5/Source/Guidelines/fr/BIB.xml
	   A /trunk/P5/Source/Guidelines/fr/CC.xml
	   A /trunk/P5/Source/Guidelines/fr/CE.xml
	   A /trunk/P5/Source/Guidelines/fr/CF.xml
	   A /trunk/P5/Source/Guidelines/fr/CH.xml
	   A /trunk/P5/Source/Guidelines/fr/CHfig.png
	   A /trunk/P5/Source/Guidelines/fr/CO.xml
	   A /trunk/P5/Source/Guidelines/fr/COL.xml
	   A /trunk/P5/Source/Guidelines/fr/DEDICATION.xml
	   A /trunk/P5/Source/Guidelines/fr/DI.xml
	   A /trunk/P5/Source/Guidelines/fr/DR.xml
	   A /trunk/P5/Source/Guidelines/fr/DS.xml
	   A /trunk/P5/Source/Guidelines/fr/DT.xml
	   A /trunk/P5/Source/Guidelines/fr/FD.xml
	   A /trunk/P5/Source/Guidelines/fr/FM1.xml
	   A /trunk/P5/Source/Guidelines/fr/FS.xml
	   A /trunk/P5/Source/Guidelines/fr/FT.xml
	   A /trunk/P5/Source/Guidelines/fr/GD.xml
	   A /trunk/P5/Source/Guidelines/fr/IN.xml
	   A /trunk/P5/Source/Guidelines/fr/MD.xml
	   A /trunk/P5/Source/Guidelines/fr/MS.xml
	   A /trunk/P5/Source/Guidelines/fr/ND.xml
	   A /trunk/P5/Source/Guidelines/fr/NH.xml
	   A /trunk/P5/Source/Guidelines/fr/PARTIND.xml
	   A /trunk/P5/Source/Guidelines/fr/PH.xml
	   A /trunk/P5/Source/Guidelines/fr/PREFS.xml
	   A /trunk/P5/Source/Guidelines/fr/REFCLA.xml
	   A /trunk/P5/Source/Guidelines/fr/REFENT.xml
	   A /trunk/P5/Source/Guidelines/fr/REFTAG.xml
	   A /trunk/P5/Source/Guidelines/fr/SA.xml
	   A /trunk/P5/Source/Guidelines/fr/SG.xml
	   A /trunk/P5/Source/Guidelines/fr/SH.xml
	   A /trunk/P5/Source/Guidelines/fr/ST.xml
	   A /trunk/P5/Source/Guidelines/fr/TC.xml
	   A /trunk/P5/Source/Guidelines/fr/TD.xml
	   A /trunk/P5/Source/Guidelines/fr/TE.xml
	   A /trunk/P5/Source/Guidelines/fr/TS.xml
	   A /trunk/P5/Source/Guidelines/fr/VE.xml
	   A /trunk/P5/Source/Guidelines/fr/WD.xml
	   A /trunk/P5/Source/Guidelines/fr/driver.xml
	   A /trunk/P5/Source/Guidelines/fr/ents.dtd
	   A /trunk/P5/Source/Guidelines/fr/fileents.dtd
	   A /trunk/P5/Source/Guidelines/fr/x
	   A /trunk/P5/Source/Guidelines/fr/xmlpic.png

	put back symbolic links

2006-03-23  Sebastian Rahtz <sebastian.rahtz@oucs.ox.ac.uk>
	Changed paths:
	   D /trunk/P5/Source/Guidelines/fr/AB.xml
	   D /trunk/P5/Source/Guidelines/fr/AI.xml
	   D /trunk/P5/Source/Guidelines/fr/BIB.xml
	   D /trunk/P5/Source/Guidelines/fr/CC.xml
	   D /trunk/P5/Source/Guidelines/fr/CE.xml
	   D /trunk/P5/Source/Guidelines/fr/CF.xml
	   D /trunk/P5/Source/Guidelines/fr/CH.xml
	   D /trunk/P5/Source/Guidelines/fr/CHfig.png
	   D /trunk/P5/Source/Guidelines/fr/CO.xml
	   D /trunk/P5/Source/Guidelines/fr/COL.xml
	   D /trunk/P5/Source/Guidelines/fr/DEDICATION.xml
	   D /trunk/P5/Source/Guidelines/fr/DI.xml
	   D /trunk/P5/Source/Guidelines/fr/DR.xml
	   D /trunk/P5/Source/Guidelines/fr/DS.xml
	   D /trunk/P5/Source/Guidelines/fr/DT.xml
	   D /trunk/P5/Source/Guidelines/fr/FD.xml
	   D /trunk/P5/Source/Guidelines/fr/FM1.xml
	   D /trunk/P5/Source/Guidelines/fr/FS.xml
	   D /trunk/P5/Source/Guidelines/fr/FT.xml
	   D /trunk/P5/Source/Guidelines/fr/GD.xml

Roma/roma.sh
#!/bin/bash
#
# Roma: transform TEI ODD files to schema and documentation
#
# Sebastian Rahtz, April 2005
# copyright: TEI Consortium
# license: GPL
# $Id$
#

makeODD() 
{
    echo "1. expand and simplify ODD "
    if test "x$lang" = "x"
    then
	xmllint --noent --xinclude $ODD \
	    | xsltproc -o $RESULTS/$ODD.compiled \
	    $SELECTEDSCHEMA $LANGUAGE $DOCLANG \
	    --stringparam useVersionFromTEI $useVersionFromTEI \
	    --stringparam TEIC $TEIC \
	    --stringparam TEISERVER $TEISERVER  \
	    --stringparam localsource "$LOCAL"  \
	    $DEBUG  $TEIXSLDIR/odds/odd2odd.xsl -
    else
	echo  [names translated to language $lang]
	xmllint --noent --xinclude $ODD \
	    | xsltproc \
	    --stringparam TEIC $TEIC \
	    --stringparam useVersionFromTEI $useVersionFromTEI \
	    --stringparam TEISERVER $TEISERVER  \
	    --stringparam localsource "$LOCAL"  \
	   $DEBUG  $TEIXSLDIR/odds/odd2odd.xsl - \
	    | xsltproc -o $RESULTS/$ODD.compiled $DEBUG $LANGUAGE $DOCLANG --stringparam TEISERVER $TEISERVER  \
	    $TEIXSLDIR/odds/translate-odd.xsl - 
    fi
}

makeRelax() 
{
    echo "2. make RELAX NG from compiled ODD"
    xsltproc $PATTERN $DEBUG $LANGUAGE $DOCLANG  $SELECTEDSCHEMA \
	     --stringparam parameterize $parameterize \
             --stringparam TEIC $TEIC \
             --stringparam outputDir $RESULTS   \
             $TEIXSLDIR/odds/odd2relax.xsl $RESULTS/$ODD.compiled
    (cd $RESULTS; \
    echo "3. make RELAX NG compact from XML"; \
    trang $schema.rng $schema.rnc  || die " trang conversion to RNC fails"; \
    xmllint --format $schema.rng > $$.xml; mv $$.xml $schema.rng )
 }

makeXSD()
{
    echo "4. make XSD from RELAX NG"
    (cd $RESULTS; \
    trang  -o disable-abstract-elements $schema.rng $schema.xsd || die " trang fails";\
    test -f xml.xsd && perl -p -i -e 's+\"xml.xsd\"+\"http://www.w3.org/2004/10/xml.xsd\"+' $schema.xsd)
}

makeDTD()
{
    echo "5. make DTD from compiled ODD"
    xsltproc  $DEBUG $LANGUAGE $DOCLANG   $SELECTEDSCHEMA \
	    --stringparam parameterize $parameterize \
	    --stringparam TEIC $TEIC \
            --stringparam outputDir $RESULTS       \
            $TEIXSLDIR/odds/odd2dtd.xsl $RESULTS/$ODD.compiled
}

makeSCH()
{
    echo "8. extract ISO Schematron from compiled ODD"
    xsltproc  $DEBUG $LANGUAGE $DOCLANG   $SELECTEDSCHEMA \
            $TEIXSLDIR/odds/extract-sch.xsl $RESULTS/$ODD.compiled > $RESULTS/$schema.sch
}

makeHTMLDOC() 
{
    echo "9. make HTML documentation $schema.doc.html "
    xsltproc 	-o $RESULTS/$schema.doc.html \
	$DEBUG  $LANGUAGE $DOCLANG --stringparam TEIC $TEIC \
	--stringparam STDOUT true \
	--stringparam splitLevel -1 \
	$DOCFLAGS $TEIXSLDIR/odds/odd2html.xsl $RESULTS/$ODD.compiled
}

makePDFDOC() 
{
    echo "7. make PDF documentation $schema.doc.pdf and $schema.doc.tex "
    xsltproc $DEBUG $LANGUAGE $DOCLANG --stringparam TEIC $TEIC \
	-o $RESULTS/$schema.doc.tex \
	$TEIXSLDIR/latex/tei.xsl $RESULTS/$schema.doc.xml
    (cd $RESULTS; pdflatex $schema.doc.tex)
}

makeXMLDOC() 
{
    echo "6. make expanded documented ODD $schema.doc.xml "
    xsltproc $DEBUG $LANGUAGE $DOCLANG --stringparam TEISERVER $TEISERVER  \
	--stringparam localsource "$LOCAL"  \
	--stringparam TEIC $TEIC \
	-o $RESULTS/$schema.doc.xml \
	$TEIXSLDIR/odds/odd2lite.xsl $RESULTS/$ODD.compiled 
}


die()
{
    echo; echo
    echo "ERROR: $@."
    D=`date "+%Y-%m-%d %H:%M:%S.%N"`
    echo "This was a fatal error. $D"
    exit 1
}

usageMsg()
{
echo "Roma -- reads in a TEI ODD file that specifies a schema, and tangles it"
echo "into RelaxNG schema, DTD, and W3C XML Schema, and can weave it into an"
echo "expanded, self-documented single ODD file. "
echo "  Usage: roma [options] schemaspec [output_directory]"
echo "  options, shown with defaults:"

echo "  --xsl=$TEIXSLDIR"
echo "  --teiserver=$TEISERVER"
echo "  --localsource=$LOCALSOURCE  # local copy of P5 sources"
echo "  options, binary switches:"
echo "  --compile          # create compiled file odd"
echo "  --debug            # leave temporary files, etc."
echo "  --doc              # create expanded documented ODD (TEI Lite XML)"
echo "  --dochtml          # create HTML version of doc"
echo "  --doclang=LANG     # language for documentation"
echo "  --docpdf           # create PDF version of doc"
echo "  --lang=LANG        # language for names of attributes and elements"
echo "  --nodtd            # suppress DTD creation"
echo "  --norelax          # suppress RELAX NG creation"
echo "  --noteic           # suppress TEI-specific features"
echo "  --noxsd            # suppress W3C XML Schema creation"
echo "  --schematron       # extract Schematron rules"
echo "  --useteiversion    # use version data from TEI P5"
echo "  --parameterize     # create parameterized DTD"
echo "  --patternprefix=STRING # prefix RELAX NG patterns with STRING"
echo "  --schema=NAME      # select named schema spec"
exit 1
}

# --------- main routine starts here --------- #
TEISERVER=http://tei.oucs.ox.ac.uk/Query/
TEIXSLDIR=/usr/share/xml/tei/stylesheet
useVersionFromTEI=true
LOCALSOURCE=
LOCAL=
TEIC=true
DOCFLAGS=
doclang=
lang=
compile=false
debug=false
dtd=true
schematron=false
relax=true
xsd=true
doc=false
docpdf=false
dochtml=false
parameterize=false
SELECTEDSCHEMA=
schema=
PATTERNPREFIX=
while test $# -gt 0; do
  case $1 in
    --compile)     compile=true;dtd=false;xsd=false;doc=false;relax=false;;
    --debug)       debug=true;;
    --doc)         doc=true;;
    --docflags=*)  DOCFLAGS=`echo $1 | sed 's/.*=//'`;;
    --dochtml)     dochtml=true;;
    --doclang=*)   doclang=`echo $1 | sed 's/.*=//'`;;
    --docpdf)      docpdf=true;;
    --lang=*)      lang=`echo $1 | sed 's/.*=//'`;;
    --localsource=*) LOCALSOURCE=`echo $1 | sed 's/.*=//'`;;
    --nodtd)       dtd=false;;
    --norelax)     relax=false;;
    --noteic)      TEIC=false;;
    --noxsd)       xsd=false;;
    --schematron)      schematron=true;;
    --useteiversion=*) useVersionFromTEI=`echo $1 | sed 's/.*=//'`;;
    --parameterize)       parameterize=true;;
    --schema=*)    schema=`echo $1 | sed 's/.*=//'`;;
    --patternprefix=*) PATTERNPREFIX=`echo $1 | sed 's/.*=//'`;;
    --teiserver=*) TEISERVER=`echo $1 | sed 's/.*=//'`;;
    --xsl=*)       TEIXSLDIR=`echo $1 | sed 's/.*=//'`;;
    --help)        usageMsg;;
     *) if test "$1" = "${1#--}" ; then 
	   break
	else
	   echo "WARNING: Unrecognized option '$1' ignored"
	   echo "For usage syntax issue $0 --help"
	fi ;;
  esac
  shift
done
ODD=${1:?"no schemaspec (i.e., ODD file) supplied; for usage syntax issue $0 --help"}
RESULTS=${2:-RomaResults}
H=`pwd`
D=`date "+%Y-%m-%d %H:%M:%S"`
echo "========= $D Roma starts, info:"
echo "Test for software: xmllint, xsltproc, trang, and perl"
which xmllint || die "you do not have xmllint"
which xsltproc || die "you do not have xsltproc"
which trang || die "you do not have trang"
which perl || die "you do not have perl"
test -f $ODD || die "file $ODD does not exist"
echo "TEI stylesheet tree: $TEIXSLDIR"
test -d $TEIXSLDIR/odds || \
     GET -e -d $TEIXSLDIR/odds/odd2odd.xsl > /dev/null || \
     die "stylesheet location $TEIXSLDIR is not accessible"
if test "x$schema" = "x"
then
 schema=$(xsltproc --xinclude $TEIXSLDIR/odds/extract-schemaSpec-ident.xsl $1 | head -1)
 schema=${schema:?"Unable to ascertain ident= of <schemaSpec>"}
fi
echo "Results to: $RESULTS"
mkdir -p $RESULTS || die "cannot make directory $RESULTS"
D=`date "+%Y-%m-%d %H:%M:%S.%N"`
echo "Process $ODD to create $schema{.dtd|.xsd|.doc.xml|.rng|.rnc} in $RESULTS"
echo "========= $D Roma starts, execution:"
if test "x$PATTERNPREFIX" = "x"
then
   PATTERN=" "
else
   PATTERN=" --stringparam patternPrefix $PATTERNPREFIX"
fi

if $debug
then
    DEBUG=" --stringparam verbose true"
else
    if $compile
	then
	DEBUG=" --stringparam stripped true"
    else
	DEBUG=" "
    fi
fi

SELECTEDSCHEMA=" --stringparam selectedSchema $schema "

if test "x$doclang" = "x"
then
  DOCLANG=" "
else 
  DOCLANG=" --stringparam doclang $doclang --stringparam documentationLanguage $doclang"
fi

if test "x$lang" = "x"
then
  LANGUAGE=" "
else 
  LANGUAGE=" --stringparam lang $lang "
fi
if test "x$LOCALSOURCE" = "x"
then
    echo using $TEISERVER to access TEI database
else
    echo using $LOCALSOURCE to gather information about TEI 
cat > subset.xsl <<EOF
<xsl:stylesheet version="1.0"
  xmlns:tei="http://www.tei-c.org/ns/1.0"
  xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
  <xsl:key name="ALL" use="1" 
    match="tei:elementSpec|tei:macroSpec|tei:classSpec|tei:moduleSpec"/>
  <xsl:template match="/">
    <tei:TEI>
      <xsl:copy-of select="tei:TEI/tei:teiHeader"/>
      <xsl:copy-of select="key('ALL',1)"/>
    </tei:TEI>
</xsl:template>
</xsl:stylesheet>
EOF
xmllint --noent --xinclude $LOCALSOURCE | xsltproc -o tei$$.xml $DEBUG subset.xsl - || die "failed to extract subset from $LOCALSOURCE "
LOCAL=$H/tei$$.xml
fi

makeODD
( $relax || $xsd ) && makeRelax
if $xsd 
then
   if ! $relax 
   then
      echo Ignored --norelax, RELAX NG required\
     to generate W3C XML Schema
   fi
   makeXSD
fi
$dtd && makeDTD
$schematron && makeSCH
$dochtml && doc=true
$docpdf && doc=true
if $doc
then 
    makeXMLDOC
    if $docpdf
    then
	makePDFDOC
    fi
fi
$dochtml && makeHTMLDOC
$compile || $debug || rm  $RESULTS/$ODD.compiled
test -f subset.xsl && rm subset.xsl
test -f tei$$.xml && rm tei$$.xml
D=`date "+%Y-%m-%d %H:%M:%S.%N"`
echo "========= $D Roma ends"
